[SourceKit/CodeFormat] Re-work and improve the indentation implementation

nathawes opened the pull request:
This restructures the indentation logic around producing a single `IndentContext`
for the line being indented. An `IndentContext` has:
- a `ContextLoc`, which points to a source location to indent relative to
- a `Kind`, indicating whether we should indent relative to that location exactly or to the start of the content on its containing line, and
- an `IndentLevel` with the relative number of levels to indent by.

It also improves the handling of:
- chained and nested parens, braces, square brackets and angle brackets, and how those interact with the exact alignment of parameters, call arguments, and tuple, array and dictionary elements.
- Indenting to the correct level after an incomplete expression, statement or decl.

Resolves:
rdar://problem/59135010
rdar://problem/25519439
rdar://problem/50137394
rdar://problem/48410444
rdar://problem/48643521
rdar://problem/42171947
rdar://problem/40130724
rdar://problem/41405163
rdar://problem/39367027
rdar://problem/36332430
rdar://problem/34464828
rdar://problem/33113738
rdar://problem/32314354
rdar://problem/30106520
rdar://problem/29773848
rdar://problem/27301544
rdar://problem/27776466
rdar://problem/27230819
rdar://problem/25490868
rdar://problem/23482354
rdar://problem/20193017
rdar://problem/47117735
rdar://problem/55950781
rdar://problem/55939440
rdar://problem/53247352
rdar://problem/54326612
rdar://problem/53131527
rdar://problem/48399673
rdar://problem/51361639
rdar://problem/58285950
rdar://problem/58286076
rdar://problem/53828204
rdar://problem/58286182
rdar://problem/58504167
rdar://problem/58286327
rdar://problem/53828026
rdar://problem/57623821
rdar://problem/56965360
rdar://problem/54470937
rdar://problem/55580761
rdar://problem/46928002
rdar://problem/35807378
rdar://problem/39397252
rdar://problem/26692035
rdar://problem/33760223
rdar://problem/48934744
rdar://problem/43315903
rdar://problem/24630624

Commit:
[SourceKit/CodeFormat] Re-work and improve the indentation implementation.

This restructures the indentation logic around producing a single IndentContext
for the line being indented. An IndentContext has:
- a ContextLoc, which points to a source location to indent relative to,
- a Kind, indicating whether we should align with that location exactly, or
  with the start of the content on its containing line, and
- an IndentLevel with the relative number of levels to indent by.

It also improves the handling of:
- chained and nested parens, braces, square brackets and angle brackets, and
  how those interact with the exact alignment of parameters, call arguments,
  and tuple, array and dictionary elements.
- Indenting to the correct level after an incomplete expression, statement or
  decl.

Resolves:
rdar://problem/59135010
rdar://problem/25519439
rdar://problem/50137394
rdar://problem/48410444
rdar://problem/48643521
rdar://problem/42171947
rdar://problem/40130724
rdar://problem/41405163
rdar://problem/39367027
rdar://problem/36332430
rdar://problem/34464828
rdar://problem/33113738
rdar://problem/32314354
rdar://problem/30106520
rdar://problem/29773848
rdar://problem/27301544
rdar://problem/27776466
rdar://problem/27230819
rdar://problem/25490868
rdar://problem/23482354
rdar://problem/20193017
rdar://problem/47117735
rdar://problem/55950781
rdar://problem/55939440
rdar://problem/53247352
rdar://problem/54326612
rdar://problem/53131527
rdar://problem/48399673
rdar://problem/51361639
rdar://problem/58285950
rdar://problem/58286076
rdar://problem/53828204
rdar://problem/58286182
rdar://problem/58504167
rdar://problem/58286327
rdar://problem/53828026
rdar://problem/57623821
rdar://problem/56965360
rdar://problem/54470937
rdar://problem/55580761
rdar://problem/46928002
rdar://problem/35807378
rdar://problem/39397252
rdar://problem/26692035
rdar://problem/33760223
rdar://problem/48934744
rdar://problem/43315903
rdar://problem/24630624

nathawes commented:
@swift-ci please test

theblixguy commented:
Wow, that’s a lot of radars!

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

Commit:
[Parse] Fix up a few places where we weren't propagating parse error status correctly.

We were always dropping the error status when returning from parseExprImpl. We
were also incorrectly keeping error status after recovering by finding the
right close token in parseList. This change fixes both, and also updates a few
callers of parseList that assumed when they reported a failure parsing an
element the list as a whole would get error status, which isn't true due to
recovery.

nathawes commented:
@swift-ci please test

rintaro commented:
@swift-ci Please test

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -568,9 +559,12 @@ ParserResult<Expr> Parser::parseExprKeyPath() {
 
   // FIXME: diagnostics
   ParserResult<Expr> rootResult, pathResult;
+  Optional<ParserStatus> parseStatus;
```
You can simply use `ParserStatus parseStatus` here. The default is success. What I was talking is that `ParserResult` cannot be success without the value. But `ParserStatus` can have simple "success" value.

nathawes commented:
@swift-ci please test

nathawes commented:
Failures seem unrelated.
@swift-ci please smoke test

nathawes commented:
@swift-ci please smoke test OS X Platform

benlangmuir reviewed (commented):
Wow, this is a big patch.  So far I've reviewed everything except the million overloads of `FormatWalker:: getIndentContextFrom`, of which I've sampled a handful of cases.  Overall, I think the approach you've taken is better and more understandable than the previous version, and I'm happy with the overall structure.  On the other hand it's not really possible to review every path here.

nathawes commented on lib/IDE/Formatting.cpp:
```diff
@@ -23,85 +24,174 @@ using namespace ide;
+
+static SourceLoc
+getLocForContentStartOnSameLine(SourceManager &SM, SourceLoc Loc,
+                                bool SkipComments = true) {
```
Oops, yeah its only call site definitely wants comments.

nathawes commented on lib/IDE/Formatting.cpp:
```diff
@@ -23,85 +24,174 @@ using namespace ide;
+    ? CommentRetentionMode::None
+    : CommentRetentionMode::ReturnAsTokens;
+  Token Given = Lexer::getTokenAtLocation(SM, Loc, Mode);
+  assert(Given.getLoc() == Loc);
```
Good point. Will do.

nathawes commented on lib/IDE/Formatting.cpp:
```diff
@@ -23,85 +24,174 @@ using namespace ide;
+  : ContextLoc(Context), Kind(Kind), IndentLevel(AddsIndent ? 1 : 0) {
+    assert(Context.isValid());
+  }
+  IndentContext(SourceLoc Context, unsigned NumIndents,
```
No sorry, it's dead code. I used to have cases where I needed more than one indent relative to the ContextLoc, but I reworked it to never need it.

nathawes commented on lib/IDE/Formatting.cpp:
```diff
@@ -23,85 +24,174 @@ using namespace ide;
+  /// Applies the overriding ContextLoc and Kind to the given IndentContext if it
+  /// starts after ApplicableFromLoc but on the same line, and its kind isn't
+  /// already 'Exact'.
+  void applyIfNeeded(SourceManager &SM, IndentContext &Ctx) {
```
I think it should actually have been an assertion. It's a case that should never happen. Context's that are 'Exact' should always set an 'Exact' context override for following child braces/parens/etc that span multiple lines to be handled correctly. If we have an 'Exact' innermost indent context with a non-exact context override  (or an exact one with a different ContextLoc) we did something wrong.

Commit:
[SourceKit/CodeFormat] Improve documentation and fix propagation of ContextLocs.

- Rename several symbols to make it clearer whether the ranges they deal with
  are open or closed.
- Add comments documenting the implementation of OutdentChecker::hasOutdent
- Fix a bug where code after a doc coment block of the '/**' style was being
  indented 1 space.
- Fix IsInStringLiteral not being set if the indent target was in a string
  segment of an interpolated multiline string.
- Update OutdentChecker::hasOutdent to propagate indent contexts from
  parent parens/brackets/braces to child parens/brackets/braces that start
  later in the same line (like FormatWalker already does). This changes the
  braces in the example below to 'inherit' a ContextLoc from their parent
  square brackets, which have a ContextLoc at 'foo'. This makes the whole
  expression be correctly considered 'outdenting':

  foo(a: "hello"
      b: "hello")[x: {
    print("hello")
  }]

nathawes commented:
@swift-ci please smoke test

nathawes commented:
It looks like the swiftpm test failure *is* related and due to the parser changes. The test parses the code below and checks the reported diagnostic for `expected ')' in expression list`

```
import PackageDescription

let package = Package(
    name: "Trivial",
    targets: [
        .target(
            name: "foo",
            dependencies: []),
)
```


Prior to my parser changes it produced:
```
/tmp/t.swift:9:1: error: expected expression in container literal
)
^
/tmp/t.swift:9:2: error: expected ')' in expression list
)
 ^
/tmp/t.swift:2:22: note: to match this opening '('
let package = Package(
                     ^
```

And after:
```
/tmp/t.swift:9:1: error: expected expression in container literal
)
^
```

I think this is because when parsing the tuple arg of `Package`, `parseExprList()` calls `parseExpr()` which calls `parseExprImpl()` which gets an error status back when it tries to parse the collection. Prior to my changes`parseExprImpl()` was dropping that error status [here](https://github.com/apple/swift/pull/30187/files#diff-3f6a161822412eea4b0e705487dd88c1L65), which meant we didn't take the true branch [here](https://github.com/apple/swift/pull/30187/files#diff-6075109c74c1e8aab51a20b38525833eR1119) in `parseExprList()` and still reported the missing right token diagnostics as well. With the error status propagating properly, the true branch is taken and they're suppressed.

Based on that it seems like the new behavior is desirable. @rintaro, do you agree? Should I just update the swiftpm test?

rintaro commented:
I agree. Let's update the spm test.

benlangmuir commented on lib/IDE/Formatting.cpp:
```diff
@@ -23,85 +24,246 @@ using namespace ide;
+    if (Value) {
+      // If the existing override applies to the same line, it's ContextLoc
+      // should override this ContextLoc - don't propagate the intermediary.
+      if (isOnSameLine(SM, Value->ApplicableFrom, NewValue.ApplicableFrom)) {
```
What does this do to your example?  E.g. if you pull bar onto the same line as foo what happens to the indentation of `d` and `fatalError`, and the closing `})`?

```
foo(a: 1, b: 2)(45, bar(c: 1,
                  d: 2) {
  fatalError()
})

nathawes replied:
It's needed for the original example too. The parens of `bar`'s argument and the parens containing those both set overrides because they span multiple lines (the previous check). When the parens of `bar`'s argument are being processed, the outer parens have already set an override, so `Value` is set. The existing override's applicable loc is its left paren which is on the same line as `NewValue`'s applicable loc - the left paren of bar's argument. If this check wasn't here with the early return, we'd effectively end up keeping the ContextLoc and Kind we had before, but we'd change the ApplicableFrom loc to  be `bar`'s left paren.  When we move on to process `bar`'s trailing closure, which has a native `ContextLoc` of `bar`, the override doesn't apply to it, because `bar` is before the `ApplicableFrom` location of the override (`bar`'s left paren). The example you gave would still indent the same, because `bar` and `foo` are on the same line, but in the original example, the trailing closure will indent relative to the line containing 'bar', giving:
```
 foo(a: 1,
     b: 2)(45, bar(c: 1,
                   d: 2) {
       fatalError()
     })
```

I think I'll restructure this code a bit to hopefully make this clearer along with some more explanation.

benlangmuir commented on lib/IDE/Formatting.cpp:
```diff
@@ -115,761 +277,2279 @@ class FormatContext {
+    if (!isInCheckRange(L, R))
+      return true;
+
+    // The CheckRange is made outdenting by any parens/braces/brackets with:
```
Thanks, this is still complicated, but the examples help a lot to get a feeling for how it works!

benlangmuir commented on lib/IDE/Formatting.cpp:
```diff
@@ -115,761 +277,2279 @@ class FormatContext {
+
+      if (TrailingTarget)
+        return None;
+      return IndentContext {ContextLoc, false};
```
When would we fall through to this?

nathawes replied:
On anything in-between the ranges handled by the above, e.g. the comment below:
```
struct Foo<T, U, V>
// FIXME: some important comment about the constraints below
where
    T: Equatable,
    U: Equatable,
    V: Codable
{
}
```

benlangmuir replied:
Ahhhh, right thanks! I forgot the ranges wouldn't necessarily be continuous.

benlangmuir commented on lib/IDE/Formatting.cpp:
```diff
@@ -115,761 +277,2279 @@ class FormatContext {
+  }
+
+  Optional<Trailing>
+  checkForTrailingTarget(SourceLoc End) {
```
I found the terminology "Trailing" hard to keep straight, since we have trailing closures, trailing where clauses, trailing targets ie. "Trailing".   Would it be correct to call this something like  `PostTokenTarget`? Maybe "trailing" is the right word, but given how overloaded it is, it would be nice if we could find the most specific term here.

nathawes replied:
The 'Token' part isn't quite right, since it can just be an empty line between two tokens. Do you mean "Trailing" specifically or "TrailingTarget" too?

Commit:
[SourceKit/CodeFormat] Simplify ContextOverride a little, and only propagate ContextLocs when necessary.

We don't actually need to set a ContextOverride unless the ContextLoc and L
paren/brace/bracket are on different lines. Combined with the fact that we
only set them if the L and R parens/braces/brackets are on different lines
to, it guarantees there will be at most one override that's applicable on
any given line, which lets us simplify the logic somewhat.

nathawes commented:
@swift-ci please test

benlangmuir reviewed (approved):
LGTM. I reviewed the most recent changes, scanned through the many specific indentation context sources, did a few more in-depth looks at specific cases in a debugger, and reviewed the new tests.  We can discuss the "Trailing" terminology, but it shouldn't block the PR.  Thanks!

nathawes commented:
This was a huge PR, so thanks a lot for going through it all @benlangmuir.
@rintaro are you ok with the current state of the parser changes?

I'll do another run of the stress tester over the full compat suite (with just the format request enabled) to make sure the last two changes didn't break anything, and do a normal source compat run too just in case the parser changes somehow affect something before I merge.

nathawes commented:
@swift-ci Please Test Source Compatibility

rintaro reviewed (approved):
Parser changes look good!
